### ISSUE TITLE
Incorporate cf-metadata mailman archive from UCAR into CF GitHub website

### ISSUE THREAD

**Issue by JonathanGregory** [cf-convention/cf-convention.github.io#222]
Dear all

The [CF UCAR email archive](http://mailman.cgd.ucar.edu/pipermail/cf-metadata) has vanished, in which our discussions occurred before we started to use GitHub. There's a broken link to it on the [discussion page](http://cfconventions.org/discussion.html). We had previously discussed that it would make sense to host the archive in GitHub. Luckily @davidhassell had downloaded it a while ago (as a plain-text file in Unix mailbox format). I have converted this to HTML [looking like this](http://www.met.rdg.ac.uk/~jonathan/CF_metadata/cf-metadata-mailman) and prepared a [pull request](https://github.com/cf-convention/cf-convention.github.io/pull/221) to include in the CF GitHub website instead of the broken link.

Does this look OK? Please note that this isn't finished yet because `faq.md` contains references to particular emails and I'll have to work out which ones they are. I have modified `discussion.md` and `governance.md`, I believe. However, this pull request contains so many files that I can't look at it. Also, the first automatic check failed. I would appreciate GitHub advice.

Best wishes

Jonathan

**Comment by davidhassell** [comment]
Dear Jonathan - I think that the new page looks fine. These new files are large (collectively at least 75 MB), but they're an essential artifact that will never change, so I think that GitHub can cope!

**Comment by erget** [comment]
The discussion pages look fine (although I've noted after downloading the page as a whole that `index.html` contains a link to `discussion.html` at root, which is fine on the website because the shoe fits, but it's not fine when you're looking at a local copy, which in my case takes me to `C:\discussion.html` and thus to nirvana).

I ended up booting my Linux in order to take a look at this. Indeed, `discussion.md` and `governance.md` have been changed. Found it out like this:

```
$ git diff --name-status main | grep -v Data/cf-metadata-mailman
M       discussion.md
M       governance.md
```

And the differences in those are:

```
$ git diff main *md
diff --git a/discussion.md b/discussion.md
index 92803cda..e5541f16 100755
--- a/discussion.md
+++ b/discussion.md
@@ -54,8 +54,7 @@ on the [CF Metadata Trac](Data/trac.html) site.
 [website]: https://cfconventions.org
 [trac]: Data/trac.html
 [about]: about-trac.html
-[mail]: http://mailman.cgd.ucar.edu/mailman/listinfo/cf-metadata
-[archives]: http://mailman.cgd.ucar.edu/pipermail/cf-metadata/
+[archives]: Data/cf-metadata-mailman
 [current]: http://cfeditor.ceda.ac.uk/proposals/1?status=active&namefilter=&proposerfilter=&descfilter=&filter+and+display=filter
 [proposals]: http://cfeditor.ceda.ac.uk/proposals/1?status=inactive&namefilter=&proposerfilter=&descfilter=&filter+and+display=filter
 [github_issues]: https://guides.github.com/features/issues
diff --git a/governance.md b/governance.md
index 19f7048b..df5ffb1a 100755
--- a/governance.md
+++ b/governance.md
@@ -83,7 +83,7 @@ The climate research community is indebted to the above for their magnanimous co

 [rules]: rules.md
 [errors]: errors.md
-[mail]: http://mailman.cgd.ucar.edu/pipermail/cf-metadata
+[mail]: Data/cf-metadata-mailman
 [html]: Data/cf-documents/cf-governance/cf2_whitepaper_final.html
 [pdf]:  Data/cf-documents/cf-governance/cf2_whitepaper_final.pdf
 [ticket146]: http://cfconventions.org/Data/Trac-tickets/146.html
```

meaning that in those files, the references to `[mail]` have been removed in the case of `discussion.md` and updated in the case of `governance.md`, and that the reference to `[archives]` had been updated. I think this is good as I've checked `discussion.md` and see no hanging references.

The only thing that remains would be the updates to `faq.md`. I believe you're probably the best person to do that - would you mind updating that as part of this PR, or would you prefer to do it in a separate issue?

**Comment by ethanrd** [comment]
Hi all - The UCAR/CGD hosted archive is back up. I touched base with CGD folks some months back and they said they were happy to continue hosting it. I just reached out again about this outage but haven't heard back yet.

**Comment by JonathanGregory** [comment]
That's useful, thanks. Shall we have both then?

**Comment by erget** [comment]
In my view, if we know that UCAR's going to host for the foreseeable future I'd prefer to hyperlink to them - it keeps our repo nice and small. If that's not the case then of course we should have our own copy.

**Comment by JonathanGregory** [comment]
I feel that it would be more robust if we had our own copy. It's helpful that UCAR CGD have hosted it for so long, and will to continue to do it, but there have been outages before, and CF isn't part of their mission. Is the size of it a problem? The HTML version of the email archive is 180 Mbyte. My copy of `cf-convention.github.io repo` is 1.1 Gbyte, the majority of which is the standard name table Data. If that's a correct reflection of the size of the repo, this isn't a huge increase, and these files will never change, so there won't be any further increase because of it. Or perhaps it could be put in a different repo of the same GitHub organisation and still be linked to the website?

**Comment by JonathanGregory** [comment]
Dear Daniel @erget et al.

http://mailman.cgd.ucar.edu/pipermail/cf-metadata has gone again this morning. I do think we would be better off if we had our own copy of this. Of course, it's helpful if UCAR keep theirs online too, since there may be other links to that site in various places. There are at present some in `faq.md`, which I can work through and replace if we decide to put our own copy online.

Above, I asked

> perhaps it could be put in a different repo of the same GitHub organisation and still be linked to the website?

I don't think that be done, can it. The other repos are not accessible to the web; only the repo with the special name ending `github.io`. If it's a problem to have the 180 Mbyte of static emails in this repo, we could create a separate GitHub organisation for it, with its own `github.io` repo. What other possibilities are there?

Best wishes

Jonathan

**Comment by davidhassell** [comment]
>  perhaps it could be put in a different repo of the same GitHub organisation and still be linked to the websit

I think we can do this: http://cfconventions.org/webtest1 is the web presence of https://github.com/cf-convention/webtest1.

So if we create a new repo called, say, `discuss-archive`for the archived discussions, we could link to it via http://cfconventions.org/discuss-archive

**Comment by erget** [comment]
I think putting it in a separate repository wouldn't be a bad idea - if we had tons of time we could always reformat the HTML pages into Markdown but I think that'd be more trouble than it's worth unless the stuff outside the body and header are static enough that we could cover them by making them part of a template. So +1 on @davidhassell 's proposed implementation of @JonathanGregory 's idea.

**Comment by JonathanGregory** [comment]
Marvellous, thanks. I misunderstood about GitHub websites. Any repo can be seen at `github.com/`_organisation_`/`_repo_ but the `github.io` one has the special URL _organisation_`.github.io`. I propose we should put the archive in a repo called `mailman`, because that's what it is.

**Comment by davidhassell** [comment]
Hello @JonathanGregory  - I'm not so sure about calling the repo `mailman`, as that's what we might call it given that it's in a particular format, not what it actually means (cf. standard names :)), hence I prefer something that indicates that it's an "archive of discussions".

**Comment by JonathanGregory** [comment]
OK - how about `mailing_list_archive`, which is what it's called on http://cfconventions.org/discussion.html?

**Comment by davidhassell** [comment]
@JonathanGregory  - `mailing_list_archive` sounds good to me.

**Comment by erget** [comment]
Fine by me. I do have 1 other idea - we could commit the archive using LFS, so it doesn't "clog" the git history - then the website would build without depending on another repo, simplifying the build process. Would that be better?

**Comment by davidhassell** [comment]
I don't fully understand the implication of using LFS, but I'm equally happy with that suggestion as I am committing them as usual.

**Comment by erget** [comment]
I fixed this by making a new PR.

```gitattributes
# Secure Jonathan's changes outside repo
$ git checkout mailman-archive
$ mv Data/cf-metadata-mailman ~
$ cp discussion.md governance.md ~
# Make new branch based on main to commit to
$ git checkout -b lfs-mailman-archive main
# Copy in data secured from Jonathan's branch
$ mv ~/cf-metadata-mailman Data/
$ cp ~/discussion.md ~/governance.md .
# Start committing that stuff
$ git add discussion.md governance.md
$ git lfs track "Data/cf-metadata-mailman/**"
Tracking "Data/cf-metadata-mailman/**"
$ git add .gitattributes
$ git add Data/cf-metadata-mailman/*
$ git commit -m "Add CF UCAR email archive"
...
$ git push erget
```

Result's in https://github.com/cf-convention/cf-convention.github.io/pull/225.

**Comment by JonathanGregory** [comment]
Thanks very much for this magic, @erget! I have closed https://github.com/cf-convention/cf-convention.github.io/pull/221.

The usual rule is that we should wait three weeks for further comments before this is merged. In this case, it's a bit inconvenient because `faq.md` needs to be updated as well, and it's difficult to work on because the pull request is too large to be viewed. I think we could merge it sooner by agreement, as if it were an update to the website. It would be helpful if anyone who has comments could make them known as soon as possible.

The proposal is to

1. Install in this GitHub repo a copy of the cf-metadata mailing list archive. We no longer use the mailing list but the archive of postings is valuable information. It is still hosted at UCAR but currently offline. Daniel @erget has made a PR which adds the archive (180 Mbyte) to the repo with GitHub large file support, so that it's not tracked. It's a static blob.

2. Change the website to point to the GitHub copy of the archive as the primary source for reference, rather than the UCAR original.

Does anyone have concerns about this? Thanks. Jonathan

**Comment by ethanrd** [comment]
Hi @JonathanGregory and all - Just two things.

I would like to understand the issue(s) impacting the NCAR/CGD archive (and give them a heads-up) before deciding to replicate cf-metadata archive on the CF Website. It is unclear to me the extent of the problem as I have only found the NCAR/CGD cf-metadata archive site inaccessible once in the last few days. Are others seeing this more regularly?

I'm a bit unclear on how the LFS storage works with GitHub Pages sites. Does storing the email archive in LFS mean these files will not impact the size of the published CF website? The [size limit](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits) for published GitHub Pages sites is 1 GB and the current build artifact is 865 MB. So, we are somewhat close to that limit. In the past, we've discussed other ways to reduce the size of the site. It might be worth revisiting that discussion at some point.

**Comment by JonathanGregory** [comment]
Dear @ethanrd 

I don't think the UCAR archive goes offline "regularly" but it has happened more than once before. (It's still offline now.) But even if it were perfectly reliable, wouldn't it be better for CF to host its own documents?

I don't know the rules about LFS. If it's problematic, other suggested alternatives are (1) put it in a different repo of our GitHub organisation. Then it would be under `github.com` instead of `github.io` - would it count as the same site? (2) put it in a different GitHub organisation also managed by us.

Best wishes

Jonathan

**Comment by JonathanGregory** [comment]
Dear all

I have been looking around for an answer to the question which @ethanrd asked

> Does storing the email archive in LFS mean these files will not impact the size of the published CF website? The [size limit]( https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits) for published GitHub Pages sites is 1 GB

I can't find a direct answer to this question, but https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage says (at the bottom)

> Git LFS cannot be used with GitHub Pages sites

I assume this means that, although we can put the files in the repo with LFS to avoid it being counted in the repo size, we can't access them on the web. If so, LFS isn't a solution.

The other solution we discussed is to put them into another repo of the organisation, as a GitHub pages project site, such as David has demonstrated. The size limits apply to repos, not to the organisation, so I suppose that would work, wouldn't it?

Cheers

Jonathan

**Comment by erget** [comment]
Hi @ethanrd @JonathanGregory , I'm not sure about the use of the Unidata archive vs. one we own ourselves - wouldn't want to tread on the toes of those who've invested time and money in maintaining the mailman server. I wasn't aware that we can expect it to continue working.

On a technical level, having a second repository is a bit cumbersome but if we expect it to be static forever I suppose it's ok, then it won't impact the build process. It's a cross-domain link but we have the fabulous link checker now so we#ll probably catch it if something goes sour. I wasn't aware about LFS not being compatible with GitHub Pages, that's a pity.

**Comment by JonathanGregory** [comment]
Dear @erget @ethanrd 

I don't see why UCAR would be offended if we have our own copy of our own material, and link to both copies from our website. That seems sensible to me just for greater resilience - wouldn't you agree? It's not surprising if things go wrong from time to time. I note that their copy is still offline. I'm grateful to them for hosting for so long. Even before we stopped actively using the mailing list, there were difficulties from time to time and Brian Eaton kindly intervened to get it working again a few times.

Best wishes

Jonathan

**Comment by ethanrd** [comment]
The [CGD-hosted CF email archive](https://mailman.cgd.ucar.edu/pipermail/cf-metadata/) has been accessible for me every time I've tried except for one instance when this issue started on 27 July. Are others still unable to access it?

**Comment by davidhassell** [comment]
Hi - I can not access it just now (2022-08-02 14:45Z). 

I was using the link http://mailman.cgd.ucar.edu/pipermail/cf-metadata/ (found from https://cfconventions.org/discussion.html) and got "_404 Not Found The requested URL /pipermail/cf-metadata/ was not found on this server._"

Is that the correct URL?

**Comment by davidhassell** [comment]
Oh, hang on! at 2022-08-02 14:47Z it now works ...

**Comment by ethanrd** [comment]
I wonder if it has something to do with using HTTP rather than HTTPS. I believe all UCAR sites are required to only support HTTPS. Though usually HTTP requests get forwarded to the matching HTTPS request.

**Comment by JonathanGregory** [comment]
I think you're right that it must be something to do with http vs https, @ethanrd, thanks. It didn't work for me with http just now. Then I tried https, and now it works with both. Good. So it _is_ working, and we can of course change https://cfconventions.org/discussion.html to use https.

But still, I think we should keep our own copy too, for resilience and because it seems logical to me that CF GitHub should host CF archives. Jonathan

**Comment by ethanrd** [comment]
OK. I talked with the NCAR/CGD folks involved in the current archives. They agree that a second copy would be good.
Two questions:

1. The file names and directory structure in the PR don't match the file names and structure in the current archive. Should they? Pro: Matching names/structure would make existing references to archive messages easy to update. Con: We'd have to grab another copy. I could ask folks at CGD who may be able to grab a copy from file rather than us scraping the site.
2. Do we need to use LFS? The individual files are not large. There are just lots of them, 9 or 10 thousand. (Did we decide not to keep them in a different repo?)

**Comment by JonathanGregory** [comment]
Dear @ethanrd

Thanks for asking NCAR/CGD.

1. Yes, I agree with the Pro. With the different organisation I will need to update links in the FAQ. Another Con is that the old organisation is perhaps less convenient for searching, because each year is in its own file.
2. LFS can't be presented to the web, so we can't use that, although it was a good idea. The concern was the size of the email archive altogether, which is 180 Mbyte, bringing us nearer to the GitHub pages limit. Hence a new repo in our organisation might be best. That would be fine to link from the website.

Best wishes

Jonathan

**Comment by ethanrd** [comment]
Hi @JonathanGregory, all - The CF email archive is down again and I haven't heard when/if it will be coming back up. It also has Google search indexing blocked which I think we need (I know that's how I have generally searched the archive).

So I think we should proceed with the structure you have (I think I was the last one hesitating on duplicating the archive).

From a quick review, looks like `mailing_list_archive` was the last suggestion for the name of a new repo.

Are there any other issues or decisions to make before proceeding with that plan?

**Comment by erget** [comment]
I don't believe so, the only suggestion I'd make is to name the repo with slug spacing, i.e. `mailing-list-archive` rather than with underscores because I find it easier to type. I care about the differentiation a lot less than whether there's milk in my coffee though, so I'm happy to be ignored here if someone like underscores more.

**Comment by JonathanGregory** [comment]
Dear all

As well as resilience and hosting our own archve which can be googled, there's a further advantage of offering different views, in that the file I made with David's help is all years together, whereas the NCAR/CGD one is organised by years.

We already have hyphens in other repo names, but no underscores, so hyphens are the right choice. @davidhassell, if you agree with this, please would you be so kind as to create a `mailing-list-archive` repo and give me any permission I might need to submit a pull request to it?

For future reference, @erget, do you like milk in your coffee or not? I like milk in my tea, but not sugar, thanks.

Cheers

Jonathan

**Comment by davidhassell** [comment]
Repo created: https://github.com/cf-convention/mailing-list-archive. It is public, so PRs can be submitted by anyone.

**Comment by erget** [comment]
Hi @JonathanGregory , I take mine unembellished in any way, but if I drink tea I tend to follow your example :)

So I think the next things to do are
1. Now that we have the files in the repo, get it setup with a website, 
2. Hookup the cfconventions.org website to refer to the new repo's website, and
3. Document this all somewhere so we don't forget about it because it's not super super obvious

Does that sound about right? Perhaps a topic for the upcoming hackathons?

**Comment by davidhassell** [comment]
I've just enabled the website for `mailing-list-archive` http://cfconventions.org/mailing-list-archive. It is currently just displaying the repo's minimal README.

**Comment by JonathanGregory** [comment]
Thanks, @davidhassell. I have created [PR #1](https://github.com/cf-convention/mailing-list-archive/pull/1) in the new repo to add a copy of the hypermail version of the mail archive as a directory `Data` in the repo. This is the same contents as the [copy I put up on my own website](http://www.met.rdg.ac.uk/~jonathan/CF_metadata/cf-metadata-mailman).  As before, you can't see it properly in the PR because there are too many files. Please could someone check and merge it? Once that is done, I will make a pull request in this repo (a modified version of my [earlier PR](https://github.com/cf-convention/cf-convention.github.io/pull/221)) to add a link from the [discussion page](http://cfconventions.org/discussion.html), and also correct the link to the NCAR archive to use `https` (fixing the error Ethan @ethanrd noticed).

**Comment by davidhassell** [comment]
Dear Jonathan - thanks. The view on your website looks good (screenshot below). I wonder if we could add a slightly more informative title, above the "cf-metadata-mailman by ..." line? I'd suggest "_Archive of all CF discussions from 2006 to 2019, including questions, standard names and enhancements._", simply since that's what I wrote in the [README](https://github.com/cf-convention/mailing-list-archive/blob/main/README.md), but it wouldn't have to be that!
![image](https://user-images.githubusercontent.com/8126576/187538364-f5d0e612-74ce-40a3-aa15-c98d0854e00b.png)

**Comment by erget** [comment]
> Thanks, @davidhassell. I have created [PR #1](https://github.com/cf-convention/mailing-list-archive/pull/1) in the new repo to add a copy of the hypermail version of the mail archive as a directory `Data` in the repo. This is the same contents as the [copy I put up on my own website](http://www.met.rdg.ac.uk/~jonathan/CF_metadata/cf-metadata-mailman). As before, you can't see it properly in the PR because there are too many files. Please could someone check and merge it? Once that is done, I will make a pull request in this repo (a modified version of my [earlier PR](https://github.com/cf-convention/cf-convention.github.io/pull/221)) to add a link from the [discussion page](http://cfconventions.org/discussion.html), and also correct the link to the NCAR archive to use `https` (fixing the error Ethan @ethanrd noticed).

Good approach @JonathanGregory . I've checked https://github.com/cf-convention/mailing-list-archive/pull/1 and would approve it but I don't have write access to the repo! @davidhassell can you help? I think it's because [I'm a member of the CF organisation but not an owner](https://github.com/orgs/cf-convention/people?query=role%3Aowner), so I don't get write access by default. We might change that or live with it - either way I've commented the PR so it's clear that it's been reviewed (at least to confirm that its contents have filenames that are expected, I didn't dig deeper than that on this).

**Comment by davidhassell** [comment]
Hi Daniel and Jonathan, I've given you both write permission to the repo, so you can approve and merge  - but I'll put my title suggestion in a suggest-changes review for traceability.

**Comment by erget** [comment]
@davidhassell @JonathanGregory I've updated the title in line in the PR with [David's proposal](https://github.com/cf-convention/cf-convention.github.io/issues/222#issuecomment-1232146213) - if you're happy, I'm happy with it being merged.

**Comment by JonathanGregory** [comment]
Dear @davidhassell and Daniel @erget, Thanks for your interventions. I agree about the title. I had not considered that! It's the name of the mailbox fle at the time I processed it. I'll see what I can do with manual editing. J

**Comment by davidhassell** [comment]
Good point, @ethanrd  - I have added teams permissions as you suggest, and also `info-mgt` (following the leas of other repos).

**Comment by JonathanGregory** [comment]
Dear @davidhassell @erget

Following your conversation in [PR #1](https://github.com/cf-convention/mailing-list-archive/pull/1), I edited my local copies of the html to replace the text `cf-mailman-archive`. In most places I put `mailing-list-archive` as a short form, and in the titles of the indexes I put _Archive of CF discussions from 2006 to 2019 on the cf-metadata mailing list_. You can see what it looks like again on [my website](http://www.met.rdg.ac.uk/~jonathan/CF_metadata/mailing-list-archive/index.html). I have made a [new PR](https://github.com/cf-convention/mailing-list-archive/pull/2) for this version. Is that all OK?

Jonathan

**Comment by davidhassell** [comment]
Looks good to me: 
![image](https://user-images.githubusercontent.com/8126576/187875269-2b3f8830-e5c4-4439-90b7-85889ce8e051.png)

![image](https://user-images.githubusercontent.com/8126576/187875129-d56dc216-d4df-4ddc-a2f0-b5d6df341c9e.png)

**Comment by JonathanGregory** [comment]
Good, thanks for checking, @davidhassell. I have closed the [old PR](https://github.com/cf-convention/mailing-list-archive/pull/1). Please could someone merge the [new one](https://github.com/cf-convention/mailing-list-archive/pull/2)?

**Comment by erget** [comment]
Done. So glancing back up, we've got the material in https://github.com/cf-convention/mailing-list-archive, but I'm still not clear on how to do the rest of the setup:

> 1. Now that we have the files in the repo, get it setup with a website,
> 2. Hookup the cfconventions.org website to refer to the new repo's website, and
> 3. Document this all somewhere so we don't forget about it because it's not super super obvious

(1) is done.

(2) looks like it's good to go if the hyperlink points to the right place, but currently it's just the README, whereas I believe it should point to [the archive index](https://cfconventions.org/mailing-list-archive/Data/index.html), and we could drop the reference to the README (or link to the archive index from the README).

And then (3) is still outstanding.

I'm happy to take these on in Santander, or to provide support from the sidelines if there's a volunteer now - I'm sadly a bit strapped for time in the interim. Any takers? If not I'll volunteer in 2w.

**Comment by JonathanGregory** [comment]
Thanks for merging, Daniel @erget. The mailing list archive can now be seen on the web at `https://cfconventions.org/mailing-list-archive/Data/`. I will make another PR to link it to the CF website.

**Comment by JonathanGregory** [comment]
Please could someone check and merge [PR #244](https://github.com/cf-convention/cf-convention.github.io/pull/244) to update the link from the Discussion page to point to the GitHub copy of the mailing list archive. The NCAR archive is linked as well, now with `https`, but it's not working at present. Thanks. Jonathan

**Comment by erget** [comment]
Looks good @JonathanGregory , many thanks. Merged.

@davidhassell how did you get the archive repo to be found under the cfconventions.org domain? Are there any updates to the docs that we should complete before closing this issue?

**Comment by davidhassell** [comment]
> David and I can record how we made the mailing list archive. I think the right place to record this would be in README.md of the new repo - do you agree?

That's a nice idea, and I agree about the location. I'm not sure that I have much to provide, apart from that the source data came from a simple text concatenation of each year's archive file downloaded from the NCAR site. Do remind me if I did more ...

**Comment by erget** [comment]
@davidhassell the only thing that's missing from my side is what magic makes https://cfconventions.org/mailing-list-archive/Data/ point to the data found in https://github.com/cf-convention/mailing-list-archive/tree/main/Data. I don't see an entry for `mailing-list-archive/Data/` in https://github.com/cf-convention/cf-convention.github.io - I'm sure there's a clever way of establishing the link, but how that works is a mystery to me.

**Comment by JonathanGregory** [comment]
@erget, I don't think this is anything special. The mailing list archive is a GitHub Pages project website. The URL of a project website is _organisation_`.github.io/`_repo_, which in this case is `cf-convention.github.io/mailing-list-archive`. Because `cfconventions.org` points to `cf-convention.github.io`, `cfconventions.org/mailing-list-archive/Data` points to `cf-convention.github.io/mailing-list-archive/Data`. Ta-da! I made this link yesterday without any magic.

**Comment by erget** [comment]
Wow, this is what revelation feels like... Great, then let's close this and thank you for helping me grasp that!
